decorators_library/decorators.py

```diff
@@ -1 +1,107 @@
-# implement your decorators here.
+import logging
+import signal
+import warnings
+
+from .exceptions import TimeoutError
+
+
+class debug(object):
+    
+    def __init__(self, logger=None):
+        self.logger = logger
+
+    def __call__(self, func):
+        if not self.logger:
+            logging.basicConfig()
+            log = logging.getLogger(func.__module__)
+            log.setLevel(logging.DEBUG)
+        else:
+            log = self.logger
+        
+        def func_wrapper(*args, **kwargs):
+            startmessage = 'Executing "{}" with params: {}, {}'.format(func.__name__, args, kwargs)
+            log.debug(startmessage)
+            results = func(*args, **kwargs)
+            endmessage = 'Finished "{}" execution with result: {}'.format(func.__name__, results)
+            log.debug(endmessage)
+            return results
+        return func_wrapper
+
+def timeout(time):
+    def wrapper(func):
+        def handler(signum, frame):
+            raise TimeoutError('Function call timed out')        
+        
+        def run_in_timer():
+            signal.signal(signal.SIGALRM, handler)
+            signal.alarm(time)
+            func()
+            signal.alarm(0)
+        return run_in_timer
+    return wrapper
+
+
+class count_calls(object):
+    # We init this dict outside of __init__ so static methods can access it.
+    count_dict = {}
+    
+    def __init__(self, func):
+        self.func = func
+        self.count = 0
+        # Now we pass in our self object, into the dict, so we can locate them in our static methods.
+        count_calls.count_dict[func] = self 
+
+    def __call__(self, *args, **kwargs):
+        self.count += 1
+        count_calls.count_dict[self.func].count = self.count # store updated result in the class dict
```

Comment:
If I'm not mistaken, this is essentially saying "self.count = self.count", since the corresponding value to self.func should be just self. Your code should work without that line, since you mutated self already.

I can kind of see the advantage of saving the whole object in the dictionary, it should be great if your objects have a lot of attributes. But in this case you only need the count, so I think it's not necessary.